src/Week7/JoinList.hs

```diff
@@ -0,0 +1,190 @@
+Use the Sized type class to write the following functions. 1. Implement the function
+
+indexJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> Maybe a
+
+indexJ finds the JoinList element at the specified index. 
+If the index is out of bounds, the function returns Nothing. By an index in a JoinList we mean the index in the list that it represents. 
+That is, consider a safe list indexing function
+
+(!!?) :: [a] -> Int -> Maybe a [] 
+!!? _ = Nothing
+!!? i | i < 0 = Nothing 
+(x:xs) !!? 0 = Just x 
+(x:xs) !!? i = xs !!? (i-1)
+
+which returns Just the ith element in a list (starting at zero) if such an element exists, or Nothing otherwise. 
+We also consider an updated function for converting join-lists into lists, just like jlbToList but ignoring the monoidal annotations:
+jlToList :: JoinList m a -> [a]
+jlToList Empty = []
+jlToList (Single _ a) = [a]
+jlToList (Append _ l1 l2) = jlToList l1 ++ jlToList l2
+
+We can now specify the desired behavior of indexJ. For any index i and join-list jl, it should be the case that
+   (indexJ i jl) == (jlToList jl !!? i)
+That is, calling indexJ on a join-list is the same as first convert- ing the join-list to a list and then indexing into the list. 
+The point, of course, is that indexJ can be more efficient (O(log n) versus O(n), assuming a balanced join-list), because it gets to use the size annotations to throw away whole parts of the tree at once, whereas the list indexing operation has to walk over every element.
+
+2. Implement the function
+
+dropJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+
+The dropJ function drops the first n elements from a JoinList. 
+This is analogous to the standard drop function on lists. Formally, dropJ should behave in such a way that
+jlToList (dropJ n jl) == drop n (jlToList jl).
+
+3. Finally, implement the function takeJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The takeJ function returns the first n elements of a JoinList, dropping all other elements. 
+Again, this function works similarly to the standard library take function; that is, it should be the case that
+jlToList (takeJ n jl) == take n (jlToList jl).
+Ensure that your function definitions use the size function from the Sized type class to make smart decisions about how to descend into the JoinList tree.
+
+-}
+
+
+indexJ :: (Sized b, Monoid b) => Int -> JoinList b a -> Maybe a
+indexJ _ Empty  = Nothing
+indexJ 0 (Single _ a) = Just a
+indexJ i _
+   | i < 0 = Nothing
+indexJ i (Append _ jl1 jl2)
+   | i < jl1_size = indexJ i jl1
+   | i > jl1_size + jl2_size = Nothing
+   | otherwise    = indexJ (i - jl1_size) jl2
+   where
+      jl1_size = getSize . size . tag $ jl1
+      jl2_size = getSize . size . tag $ jl2
+
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ _ Empty = Empty
+dropJ 0 jl = jl
+dropJ n (Append _ jl1 jl2)
+   | n < jl1_size  = (dropJ n jl1) +++ jl2
+   | n > jl1_size + jl2_size = Empty
+   | n == jl1_size = jl2
```

Comment:
These cases could be ordered more meaningfully, as so: 
- `n < jl1_size1` (again, don't use snake_case) 
- `n == jl1_size1`
- `jl1_size < n && n < jl2_size`
- `otherwise` (the `Empty` case should be captured here) .